java/arcs/sdk/jvm/HandleImpl.kt

```diff
@@ -0,0 +1,17 @@
+package arcs.sdk
+
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtOperation
+import arcs.core.crdt.internal.VersionMap
+
+abstract class HandleImpl<Data : CrdtData, Op : CrdtOperation, T>(
+    override val name: String,
+    val particle: Particle,
+    val storageProxy: StorageProxy<Data , Op, T>) : Handle {
```

Comment:
Nit: put the closing paren on a new line:

```
    val storageProxy: StorageProxy<Data , Op, T>
) : Handle {
```

Reply:
This sets up a cyclical relation Handle <-> Particle, I think this was a mistake in the TS runtime, and in the new architecture, the ArcHost should be responsible for propagating onParticleUpdate from the Handle to the Particle. I don't think the Handle should have a direct reference to the particle, indeed, you should be able to use the Handle class without particles. 

e.g. ArcHost creates Handles and Particles, listens for changes and invokes Particle lifecycle methods.

Reply:
Additional question for @shans which is, why do Handle's now refer to StorageProxy type, instead of Store.  The old Handle would reference Store, and StorageProxy implemented Store, so a Handle could even set directly on top of a Store (no proxy), or have a Proxy underneath. The NG Handle specifically always have a StorageProxy.  What was the rational for not using the more abstract case (Store)?

Reply:
I've been curious about another approach:
* One, shared Handle & StorageProxy implementation hierarchy used by both JVM and wasm
* In wasm-space, a storage communication implementation that converts ProxyMessage and ProxyCallback messages between `ConsumerData` and `ByteArray`

Reply:
I think Handle and StorageProxy definitely should be pure Kotlin code with no dependencies on JVM or Android APIs.  But my main issue is Handle/StorageProxy should be usable without Particle and I don't think they should directly call onParticleUpdate(), I think subscribing to updates should be an API, and the idiomatic API for particle developers can do the subscription and call lifecycle methods like onParticleUpdate() without the Handle/Proxy classes directly having to know about Particles. 

Whoever instantiates the Handle/Proxy can subscribe for callbacks, that can either be the ArcHost, or Particles themselves. e.g. an API like proxy.on { callback } or handle.on {callback} would be more abstract with less coupling to particles. 

I mean we could land this PR as is and refactor it in followups.

Reply:
fwiw, I like your callback idea Ray. Doing it in a follow-up PR sounds good to me too.

Reply:
it would also be nice to make a compatible change in the TS code, to keep the implementations consistent.

Reply:
Sounds great to me, I can do this next (unless someone has already started on it)